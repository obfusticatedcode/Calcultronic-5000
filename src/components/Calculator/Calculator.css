.Calculator__container {
  background-color: var(--calculator-background-color);
  height: var(--calculator-height);
  width: var(--calculator-width);
}