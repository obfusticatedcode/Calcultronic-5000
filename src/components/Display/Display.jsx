import React from "react";
import PropTypes from "prop-types";
import "./Display.css";

const Display = ({ displayValue }) => {
  return (
    <div className="Display__container">
      <p className="Display__value">{displayValue}</p>
    </div>
  );
};

Display.propTypes = { displayValue: PropTypes.string.isRequired };

export default Display;
