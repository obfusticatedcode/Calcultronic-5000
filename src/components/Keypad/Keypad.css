.Keypad__container {
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  height: var(--keypad-height);
  padding: 2%;
  width: var(--keypad-width);
}

.Keypad__numbersContainer {
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  height: var(--number-height);
  width: var(--number-width);
}

.Keypad__operatorsContainer {
  display: flex;
  flex-direction: column;
  height: 80%;
  width: 25%;
}

.Keypad__submitContainer {
  height: 20%;
  width: 100%;
}