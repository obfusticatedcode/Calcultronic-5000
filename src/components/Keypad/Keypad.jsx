import React from "react";
import PropTypes from "prop-types";

const Keypad = ({
  callOperator,
  numbers,
  operators,
  setOperator,
  updateDisplay
}) => {
  return <div className="Keypad__container" />;
};

Keypad.propTypes = {
  callOperator: PropTypes.func.isRequired,
  numbers: PropTypes.array.isRequired,
  operators: PropTypes.array.isRequired,
  setOperator: PropTypes.func.isRequired,
  updateDisplay: PropTypes.func.isRequired
};
export default Keypad;
