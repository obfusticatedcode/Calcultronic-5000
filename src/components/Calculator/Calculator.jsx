import React, { Component } from "react";
import Display from "../Display/Display";
import Keypad from "../Keypad/Keypad";

import "./Calculator.css";

class Calculator extends Component {
  state = {
    displayValue: "0",
    numbers: ["9", "8", "7", "6", "5", "4", "3", "2", "1", ".", "0", "ce"],
    operators: ["/", "x", "-", "+"],
    selectedOperator: "",
    storedValue: ""
  };

  callOperator = () => {
    console.log("call operation");
  };

  setOperator = value => {
    let {
      state: { displayValue, selectedOperator, storedValue }
    } = this;

    if (selectedOperator === "") {
      storedValue = displayValue;
      displayValue = "0";
      selectedOperator = value;
    } else {
      selectedOperator = value;
    }
    this.setState({ displayValue, selectedOperator, storedValue });
  };

  updateDisplay = value => {
    let {
      state: { displayValue }
    } = this;
    if (value === "." && displayValue.includes(".")) {
      value = "";
    }
    if (value === "ce") {
      displayValue = displayValue.substr(0, displayValue.length - 1);
      if (displayValue === "") {
        displayValue = "0";
      }
    } else {
      displayValue === "0" ? (displayValue = value) : (displayValue += value);
    }
    this.setState(() => ({ displayValue }));
  };

  render() {
    const {
      state: { displayValue, numbers, operators },
      callOperator,
      setOperator,
      updateDisplay
    } = this;

    return (
      <div className="Calculator__container">
        <Display displayValue={displayValue} />
        <Keypad
          callOperator={callOperator}
          numbers={numbers}
          operators={operators}
          setOperator={setOperator}
          updateDisplay={updateDisplay}
        />
      </div>
    );
  }
}

export default Calculator;
