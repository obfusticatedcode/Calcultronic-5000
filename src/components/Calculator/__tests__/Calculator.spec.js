import React from "react";
import { shallow, mount } from "enzyme";
import Calculator from "../Calculator";
import Display from "../../Display/Display";
import Keypad from "../../Keypad/Keypad";

const setup = (props = {}) => shallow(<Calculator {...props} />);
const mountedSetup = (props = {}) => mount(<Calculator {...props} />);

describe("<Calculator></Calculator>", () => {
  let wrapper = null;

  beforeEach(() => (wrapper = setup({})));
  it("should render a <div></div>", () => {
    expect(wrapper.find("div").length).toEqual(1);
  });

  it("should render <Display></Display> and <Keypad></Keypad>", () => {
    expect(
      wrapper.containsAllMatchingElements([
        <Display displayValue={wrapper.instance().state.displayValue} />,
        <Keypad
          callOperator={wrapper.instance().callOperator}
          numbers={wrapper.instance().state.numbers}
          operators={wrapper.instance().state.operators}
          setOperator={wrapper.instance().setOperator}
          updateDisplay={wrapper.instance().updateDisplay}
        />
      ])
    ).toEqual(true);
  });

  it("should render correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });
});

describe("mounted <Calculator/>", () => {
  let wrapper = null;
  beforeEach(() => (wrapper = mountedSetup({})));

  it("should call updateDisplay() when a number key is clicked", () => {
    const spy = jest.spyOn(wrapper.instance(), "updateDisplay");
    wrapper.instance().forceUpdate();
    expect(spy).toHaveBeenCalledTimes(0);
    wrapper
      .find(".Key__numberKey")
      .first()
      .simulate("click");
    expect(spy).toHaveBeenCalledTimes(1);
  });

  it("should call setOperator() when a number key is clicked", () => {
    const spy = jest.spyOn(wrapper.instance(), "setOperator");
    wrapper.instance().forceUpdate();
    expect(spy).toHaveBeenCalledTimes(0);
    wrapper
      .find(".Key__operatorKey")
      .first()
      .simulate("click");
    expect(spy).toHaveBeenCalledTimes(1);
  });

  it("should call updateDisplay() when a number key is clicked", () => {
    const spy = jest.spyOn(wrapper.instance(), "updateDisplay");
    wrapper.instance().forceUpdate();
    expect(spy).toHaveBeenCalledTimes(0);
    wrapper
      .find(".Key__numberKey")
      .first()
      .simulate("click");
    expect(spy).toHaveBeenCalledTimes(1);
  });

  it("should call callOperator() when a number key is clicked", () => {
    const spy = jest.spyOn(wrapper.instance(), "callOperator");
    wrapper.instance().forceUpdate();
    expect(spy).toHaveBeenCalledTimes(0);
    wrapper
      .find(".Key__submitKey")
      .first()
      .simulate("click");
    expect(spy).toHaveBeenCalledTimes(1);
  });
});

describe("updateDisplay()", () => {
  let wrapper = null;
  beforeEach(() => (wrapper = setup()));
  it("should update displayValue", () => {
    wrapper.instance().updateDisplay("5");
    expect(wrapper.state("displayValue")).toEqual("5");
  });

  it("should concatenate displayValue", () => {
    wrapper.instance().updateDisplay("5");
    wrapper.instance().updateDisplay("0");
    expect(wrapper.state("displayValue")).toEqual("50");
  });

  it("should remove leading `0` from displayValue", () => {
    wrapper.instance().updateDisplay("0");
    expect(wrapper.state("displayValue")).toEqual("0");
    wrapper.instance().updateDisplay("5");
    expect(wrapper.state("displayValue")).toEqual("5");
  });

  it("should prevent multiple leading `0`s from displayValue", () => {
    wrapper.instance().updateDisplay("0");
    wrapper.instance().updateDisplay("0");
    expect(wrapper.state("displayValue")).toEqual("0");
  });

  it("should remove last character from displayValue", () => {
    wrapper.instance().updateDisplay("5");
    wrapper.instance().updateDisplay("0");
    wrapper.instance().updateDisplay("ce");
    expect(wrapper.state("displayValue")).toEqual("5");
  });

  it("should prevent multiple instances of `.` in displayValue", () => {
    wrapper.instance().updateDisplay(".");
    wrapper.instance().updateDisplay(".");
    expect(wrapper.state("displayValue")).toEqual(".");
  });

  it("will set displayValue to `0` if displayValue is equal to an empty string", () => {
    wrapper.instance().updateDisplay("ce");
    expect(wrapper.state("displayValue")).toEqual("0");
  });
});

describe("setOperator", () => {
  let wrapper = null;
  beforeEach(() => (wrapper = setup({})));
  it("should update the value of selectedOperator", () => {
    wrapper.instance().setOperator("+");
    expect(wrapper.state("selectedOperator")).toEqual("+");
    wrapper.instance().setOperator("/");
    expect(wrapper.state("selectedOperator")).toEqual("/");
  });

  it("should update the value of storedValue to the value of the displayValue", () => {
    wrapper.setState({ displayValue: "5" });
    wrapper.instance().setOperator("+");
    expect(wrapper.state("storedValue")).toEqual("5");
  });

  it("selectedOperator is NOT an empty string, does not update storedValue", () => {
    wrapper.setState({ displayValue: "5" });
    wrapper.instance().setOperator("+");
    expect(wrapper.state("storedValue")).toEqual("5");
    wrapper.instance().setOperator("-");
    expect(wrapper.state("storedValue")).toEqual("5");
  });
});

describe("callOperator()", () => {
  let wrapper = null;
  beforeEach(() => (wrapper = setup({})));
  it("it should update displayValue to the sum of storedValue and displayValue", () => {
    wrapper.setState({ storedValue: "3" });
    wrapper.setState({ displayValue: "2" });
    wrapper.setState({ selectedOperator: "+" });
    wrapper.instance().callOperator();
    expect(wrapper.state("displayValue")).toEqual("5");
  });
  it("it should update displayValue to the difference of storedValue and displayValue", () => {
    wrapper.setState({ storedValue: "3" });
    wrapper.setState({ displayValue: "2" });
    wrapper.setState({ selectedOperator: "-" });
    wrapper.instance().callOperator();
    expect(wrapper.state("displayValue")).toEqual("1");
  });
  it("it should update displayValue to the product of storedValue and displayValue", () => {
    wrapper.setState({ storedValue: "3" });
    wrapper.setState({ displayValue: "2" });
    wrapper.setState({ selectedOperator: "x" });
    wrapper.instance().callOperator();
    expect(wrapper.state("displayValue")).toEqual("6");
  });
  it("it should update displayValue to the quotient of storedValue and displayValue", () => {
    wrapper.setState({ storedValue: "3" });
    wrapper.setState({ displayValue: "2" });
    wrapper.setState({ selectedOperator: "/" });
    wrapper.instance().callOperator();
    expect(wrapper.state("displayValue")).toEqual("1.5");
  });
  it("it should update displayValue to `0` if operation results in `Infinity`", () => {
    wrapper.setState({ storedValue: "7" });
    wrapper.setState({ displayValue: "0" });
    wrapper.setState({ selectedOperator: "/" });
    wrapper.instance().callOperator();
    expect(wrapper.state("displayValue")).toEqual("0");
  });
  it("it should update displayValue to `0` if selectedOperatror does not match cases", () => {
    wrapper.setState({ storedValue: "7" });
    wrapper.setState({ displayValue: "10" });
    wrapper.setState({ selectedOperator: "string" });
    wrapper.instance().callOperator();
    expect(wrapper.state("displayValue")).toEqual("0");
  });
  it("it should update displayValue to `0` if called with no value for storedValue or selectedOperator", () => {
    wrapper.setState({ storedValue: "" });
    wrapper.setState({ displayValue: "10" });
    wrapper.setState({ selectedOperator: "" });
    wrapper.instance().callOperator();
    expect(wrapper.state("displayValue")).toEqual("0");
  });
});
