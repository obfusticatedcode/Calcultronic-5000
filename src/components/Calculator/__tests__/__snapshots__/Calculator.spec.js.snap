// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Calculator></Calculator> should render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Calculator />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Display
          displayValue="0"
        />,
        <Keypad
          callOperator={[Function]}
          numbers={
            Array [
              "9",
              "8",
              "7",
              "6",
              "5",
              "4",
              "3",
              "2",
              "1",
              ".",
              "0",
              "ce",
            ]
          }
          operators={
            Array [
              "/",
              "x",
              "-",
              "+",
            ]
          }
          setOperator={[Function]}
          updateDisplay={[Function]}
        />,
      ],
      "className": "Calculator__container",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "displayValue": "0",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "callOperator": [Function],
          "numbers": Array [
            "9",
            "8",
            "7",
            "6",
            "5",
            "4",
            "3",
            "2",
            "1",
            ".",
            "0",
            "ce",
          ],
          "operators": Array [
            "/",
            "x",
            "-",
            "+",
          ],
          "setOperator": [Function],
          "updateDisplay": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Display
            displayValue="0"
          />,
          <Keypad
            callOperator={[Function]}
            numbers={
              Array [
                "9",
                "8",
                "7",
                "6",
                "5",
                "4",
                "3",
                "2",
                "1",
                ".",
                "0",
                "ce",
              ]
            }
            operators={
              Array [
                "/",
                "x",
                "-",
                "+",
              ]
            }
            setOperator={[Function]}
            updateDisplay={[Function]}
          />,
        ],
        "className": "Calculator__container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "displayValue": "0",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "callOperator": [Function],
            "numbers": Array [
              "9",
              "8",
              "7",
              "6",
              "5",
              "4",
              "3",
              "2",
              "1",
              ".",
              "0",
              "ce",
            ],
            "operators": Array [
              "/",
              "x",
              "-",
              "+",
            ],
            "setOperator": [Function],
            "updateDisplay": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
